### DISCUSSION TITLE
What are some of the trade-offs between CLU and Jaxline?

### DISCUSSION THREAD

**Discussion by marcvanzee** [google/flax#1143] [Q&A] [answered]
[CLU](https://pypi.org/project/clu/) and [Jaxline](https://pypi.org/project/jaxline/) are both libraries for simplifying some of the boiletplate code that exists in ML.

**Comment by marcvanzee** [comment, marked as answer]
**CLU**

- Open-sourced: https://github.com/google/CommonLoopUtils
- Doesn't (and never will) contain a training loop.  They believe that the training loop should be written and controlled by the user. 
- Contains a few utilities making your code and training loop short and readable (e.g. periodic_actions, deterministic_data).

**Jaxline**

- Open-sourced: https://github.com/deepmind/jaxline
- Focuses on the "running a synchronous step function in a loop on TPUs"
- Has all the boilerplate code for checkingpoint / logging / etc.

**Comment by kamalkraj** [comment]
Hi @marcvanzee,
Is there any documentation available for CLU, Other than the [clu_synopsis.ipynb](https://colab.research.google.com/github/google/CommonLoopUtils/blob/master/clu_synopsis.ipynb)

What is the function of the `platform` module. ? https://github.com/google/flax/blob/e30b7f5fff03df0840e7da40a9f8923aee6fb42b/examples/mnist/main.py#L54:L57

**Reply by andsteing** [reply]
Hi @kamalkraj
[clu_synopsis.ipynb](https://colab.research.google.com/github/google/CommonLoopUtils/blob/master/clu_synopsis.ipynb) covers most of the functionality and is our only documentation existing so far (other than the pydocs).

The [`platform`](https://github.com/google/CommonLoopUtils/blob/main/clu/platform/local.py) module is a bit special because "work unit" merely call `absl.logging.info()` *externally*, but interact with Google's experiment scheduling system *internally*. So this is very useful for keeping the same code internally and externally, but the module currently does not add any value externally, that's also why it's not documented better.

**Reply by JonathanRaynerMonumo** [reply]
@andsteing Any timeline on when clu might be opened for contributions or documentation will be developed? It seems that

- not much has happened in terms of documentation in 2 years since this thread
- clu is recommended in core examples in jax and flax, including "[Quickstart](https://flax.readthedocs.io/en/latest/getting_started.html#)"
- but things can't be improved, because clu doesn't accept PR's or opening issues

this doesn't seem like a good state of affairs for the ecosystem if wanting to grow jax/flax outside of Google...
